### DISCUSSION TITLE
Optimized_route with no specific end location

### DISCUSSION THREAD

**Discussion by mabilinab** [valhalla/valhalla#4839] [Q&A]
Hi!
Is it possible to use the optmized_route API with no specific end location? 
I just want a specific start location (therefore I'll put it at the top of the list), however, I don't have a specific end location, and don't want the last location of the list to be considered as the desired end location.

> Returns an optimized route stopping at each destination location exactly one time, always starting at the first location in the list and ending at the last location. This will result in a route with multiple legs.

This means that Valhalla will always try to end the route at the last location of the request list, which is not what I want here.

**Comment by kevinkreiser** [comment]
it is currently not supported however we could support it as a new option or set of options eg:

1. first and last are required to stay in their positions
2. first is required to stay in its position
3. last is required to stay in its position
4. no required position just tell me whatever traversal is best

if that sounds good to you then i think it can become a proposal for an issue and we can work on doing it. doesnt seem that difficult to me!

in the request it would be like:

```protobuf
enum OptimizationOrdering {
  both_fixed = 0;
  first_fixed = 1;
  last_fixed = 2;
  none_fixed = 3;
}

message Options {
  // ... other options before this one ...
  OptimizationOrdering optimization_ordering = 87;
}
```

**Reply by mabilinab** [reply]
Hey Kevin, sounds great! 
How can we move forward with this?

**Reply by kevinkreiser** [reply]
ill have a look at the simulated annealing code this evening to see if i cant just do it real quick.

**Reply by mabilinab** [reply]
Hey Kevin!
Do you have any update on this? Do you think it's easily feasible? Thank you!

**Reply by kevinkreiser** [reply]
yes it looks very feasible. Other than plumbing the request option through it seems we only need to make 2 small alterations to the simulated annealing code. I hadn't had time since I last mentioned this but will have time sunday evening. basic gist is:

1. update proto and pipe the options into the optimization solver
2. update the solver to look at the option when determining initial order and candidates for swaping
3. write unit tests

**Comment by kevinkreiser** [comment]
i have a pr started for this and a bunch of work on a local branch. i'll try to get finished with it this week as its just the right size thing for me with the amoutn of time i have at my disposal

**Reply by mabilinab** [reply]
Hey @kevinkreiser, do you happen to have any updates on this 🙏🏼?

**Reply by kevinkreiser** [reply]
I have it mostly done but not working. Maybe I can work on it for an hour on Tuesday morning

**Comment by mabilinab** [comment]
@kevinkreiser I've got a question related to the optimized_route API: does it take into account the context of the route at each point or does it just optimize independently of what was the previous point visited.

Let me put it differently: there might be a situation where the optimizer might recommend going from A -> B -> D -> C -> E (points are all on nearby streets). The theoretical shortest route from D to C is to start from D on the right street side and continue to C. But if you're at D while previously coming from B, you would be on the opposite side of the street and therefore it means you need to do a U-Turn at D to take the direction to C, but that U-turn might be prohibited.  

For instance, this issue happens with VROOM for instance, because the cost matrix is calculated independently of the whole route context. Therefore, it doesn't understand that going B -> D -> C means that when you're at D after growing through B -> D, you would be on the opposite heading of the route D -> C.

So my question is: does the Valhalla route_optimize API take into account this complexity?
Thanks!

**Reply by kevinkreiser** [reply]
It does not take into account this complexity. The only APIs that we have that do take into account this complexity are the map matching API and the routing API which allow you to specify the location stop types. of course for those APIs you need to know the order of locations in advance.

Now what we could do is add an option to the optimized route API that would allow you to set a penalty for uturns at stops so that when we compute the ordering it takes that into account. that would presumably fix the issue

**Reply by mabilinab** [reply]
> Now what we could do is add an option to the optimized route API that would allow you to set a penalty for uturns at stops

Yup, that could possibly work! Do you have any idea on when that could be available? Do I need to create a seperare github Issue?

Also, @nilsnolde I've seen you've pretty active on VROOM repo on this issue of u-turns at stops –– do you happen to have any input on this? 

Thanks!

**Reply by mabilinab** [reply]
Hi @kevinkreiser @nilsnolde any update or idea on how to best approach this 🙏🏼 ?

**Reply by mabilinab** [reply]
Just FYI I found a workaround: if you take a position on the side of the street (instead of the center) and use the preferred_side parameter to force a side of the street, valhalla won't make a u-turn from that start position. So seems like as long as you define the street side, you won't get u-turns in the cost matrix
